### PR TITLE
Initial dev of packet ccm 

### PR DESCRIPTION
**Pull request by ashleyschuett**
Fixes part of https://github.com/containership/cloud.provision/issues/267

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
+FROM golang:1.10.1-alpine3.7 as builder
+
+# Add build tools
+RUN apk update && \
+    apk add --no-cache git gcc musl-dev glide mercurial
```
Comment (mattkelly) [inline Dockerfile]: Why is `mercurial` required?

File: packet/cloud.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	"os"
+
+	"github.com/packethost/packngo"
+
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+const (
+	packetAuthToken string = "PACKET_AUTH_TOKEN"
```
Comment (mattkelly) [inline packet/cloud.go]: More specific names, e.g. `packetAuthTokenEnvVar` or something

File: packet/cloud.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	if token == "" {
+		return nil, fmt.Errorf("environment variable %q is required", packetAuthToken)
+	}
+
+	if project == "" {
+		return nil, fmt.Errorf("environment variable %q is required", packetProjectID)
+	}
+
+	pClient := packngo.NewClient("", token, nil)
```
Comment (mattkelly) [inline packet/cloud.go]: [Hungarian notation isn't preferred in go AFAIK](https://github.com/golang/lint/issues/258#issuecomment-260288198)

File: packet/cloud.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	facility, err := deviceFacility()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get facility from device metadata: %s", err)
```
Comment (mattkelly) [inline packet/cloud.go]: `errors.Wrap()` here and everywhere applicable

File: packet/cloud.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	}, nil
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(providerName, func(config io.Reader) (cloudprovider.Interface, error) {
+		return newCloud(config)
+	})
+}
+
+func (c *cloud) Initialize(clientBuilder controller.ControllerClientBuilder) {
```
Comment (mattkelly) [inline packet/cloud.go]: Comments on exported functions - lint please

Comment (mattkelly) [inline packet/cloud.go]: `Initialize(_ controller.ControllerClientBuilder)`

File: packet/devices.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	return nodeAddresses(device)
+}
+
+func nodeAddresses(device *packngo.Device) ([]v1.NodeAddress, error) {
+	var addresses []v1.NodeAddress
+	addresses = append(addresses, v1.NodeAddress{Type: v1.NodeHostName, Address: device.Hostname})
+
+	var privateIP, publicIP string
+	for _, address := range device.Network {
+		if address.AddressFamily == 4 {
```
Comment (mattkelly) [inline packet/devices.go]: Use [this constant](https://github.com/packethost/packngo/blob/95039464827381cb3f66f5dfb1b5ad3e18cf052b/metadata/metadata.go#L56)

File: packet/devices.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	if err == nil {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// InstanceShutdownByProviderID returns true if the instance is shutdown in cloudprovider
+//
+// TODO:// is there a hook for this?
```
Comment (mattkelly) [inline packet/devices.go]: Can we answer this before merging?

File: packet/metadata.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	"io/ioutil"
+	"net/http"
+)
+
+const (
+	packetMetadataURL = "https://metadata.packet.net/2009-04-04/meta-data"
+	facilityMetadata  = "facility"
+)
+
+// deviceFacility returns the favility of the currently device
```
Comment (mattkelly) [inline packet/metadata.go]: Spelling

File: packet/metadata.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+func deviceFacility() (string, error) {
+	return httpGet(fmt.Sprintf("%s/%s", packetMetadataURL, facilityMetadata))
+}
+
+// httpGet is for getting different metadata about the machine that is retrievable
+// from "https://metadata.packet.net/2009-04-04/meta-data"
```
Comment (mattkelly) [inline packet/metadata.go]: Maintainability: don't put the value of the constant `packetMetadataURL` here.

File: Dockerfile (hunk 1 above)
Comment (ashleyschuett) [reply Dockerfile]: the docker image wouldn't build without it. As it is now a required prereq for one of the packages

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+RUN cd $SRC_DIR && \
+    CGO_ENABLED=0 GOOS=linux GOARCH=amd64 \
+        go build \
+        -a -tags netgo \
+        -o packet-cloud-controller-manager ./ && \
+    cp packet-cloud-controller-manager /bin/
+
+# Create Docker image of just the binary
+FROM scratch as runner
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
```
[the remark this answers is not in the document]
Comment (ashleyschuett) [reply Dockerfile]: `facility from device metadata: Get https://metadata.packet.net/2009-04-04/meta-data/facility: x509: failed to load system roots and no roots provided`

File: packet/devices.go (hunk 7 above)
Comment (ashleyschuett) [reply packet/devices.go]: I wanted to ask someone as packet, as far as i can see from the api docs there is not a way to know this.

File: Gopkg.toml [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+#
+# [prune]
+#   non-go = false
+#   go-tests = true
+#   unused-packages = true
+
+[[override]]
+  # only needed until kubernetes moves to dep or github.com/docker/distribution's latest tag builds with kubernetes.
+  name = "github.com/docker/distribution"
+  #revision = "master"
```
Comment (mattkelly) [inline Gopkg.toml]: Remove

File: Gopkg.toml [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  name = "k8s.io/apiserver"
+  version = "kubernetes-1.11.0"
+
+[[constraint]]
+  name = "k8s.io/kubernetes"
+  version = "1.11.0"
+
+[prune]
+  go-tests = true
+  #unused-packages = true
```
Comment (mattkelly) [inline Gopkg.toml]: Remove maybe?

File: packet/metadata.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+// httpGet is for getting different metadata about the machine
+func httpGet(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.New(fmt.Sprintf("device metadata returned non 200 status code: %d", resp.StatusCode))
```
Comment (mattkelly) [inline packet/metadata.go]: You can use `errors.Errorf`

File: packet/cloud.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+	instances cloudprovider.Instances
+	zones     cloudprovider.Zones
+}
+
+func newCloud(config io.Reader) (cloudprovider.Interface, error) {
+	token := os.Getenv(packetAuthTokenEnvVar)
+	project := os.Getenv(packetProjectIDEnvVar)
+
+	if token == "" {
+		return nil, errors.New(fmt.Sprintf("environment variable %q is required", packetAuthTokenEnvVar))
```
Comment (mattkelly) [inline packet/cloud.go]: `Errorf`

File: packet/cloud.go [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	if project == "" {
+		return nil, errors.New(fmt.Sprintf("environment variable %q is required", packetProjectIDEnvVar))
```
Comment (mattkelly) [inline packet/cloud.go]: `Errorf`

File: packet/devices.go [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+// The providerID spec should be retrievable from the Kubernetes
+// node object. The expected format is: packet://device-id
+func deviceIDFromProviderID(providerID string) (string, error) {
+	if providerID == "" {
+		return "", errors.New("providerID cannot be empty string")
+	}
+
+	split := strings.Split(providerID, "://")
+	if len(split) != 2 {
+		return "", errors.New(fmt.Sprintf("unexpected providerID format: %s, format should be: packet://device-id", providerID))
```
Comment (mattkelly) [inline packet/devices.go]: `Errorf`

File: packet/devices.go [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	if split[0] != providerName {
+		return "", errors.New(fmt.Sprintf("provider name from providerID should be packet: %s", providerID))
```
Comment (mattkelly) [inline packet/devices.go]: `Errorf`

File: Makefile [hunk 18]
```diff
@@ -0,0 +1,40 @@
+SHELL=/bin/bash
+DOCKERREGISTRY ?= "packethost"
+CCM_IMAGE_TAG ?= "latest"
+PKG_LIST := $(shell glide novendor)
+GO_FILES := $(shell find . -type f -not -path './vendor/*' -name '*.go')
+
+.PHONY: all fmt-check lint test vet release
```
Comment (mattkelly) [inline Makefile]: `all` and `release` are not targets

File: packet/cloud.go [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+	})
+}
+
+// Initialize provides the cloud with a kubernetes client builder and may spawn goroutines
+// to perform housekeeping activities within the cloud provider.
+func (c *cloud) Initialize(_ controller.ControllerClientBuilder) {
+}
+
+// LoadBalancer returns a balancer interface. Also returns true if the interface is supported, false otherwise.
+func (c *cloud) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
```
Comment (mattkelly) [inline packet/cloud.go]: Maybe append another line to comments here saying `TODO unimplemented` or something?

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,35 @@
+FROM golang:1.10.1-alpine3.7 as builder
+
+# Add build tools
+RUN apk update && \
+    apk add --no-cache git gcc musl-dev glide mercurial && \
```
Comment (mattkelly) [inline Dockerfile]: `glide` no longer needed

**Pull request merged**